File: js/test/test_cross_language.js
/**
 * Cross-language compatibility tests for TYTX protocol.
 *
 * These tests verify that JavaScript implementation produces output
 * identical to the shared fixtures that Python and TS must also match.
 *
 * @module test_cross_language
 */

const { describe, test } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const { registry } = require('../src/registry');
require('../src/types'); // Register built-in types
const { as_typed_json, from_json } = require('../src/json_utils');

// Load fixtures
const fixturesPath = path.join(__dirname, '../../tests/fixtures/cross_language.json');
const fixtures = JSON.parse(fs.readFileSync(fixturesPath, 'utf-8'));

/**
 * Create JavaScript value from fixture spec.
 */
function createValue(spec) {
    if (spec && typeof spec === 'object' && 'type' in spec) {
        const typeName = spec.type;
        const value = spec.value;

        switch (typeName) {
            case 'int':
                return parseInt(value, 10);
            case 'float':
                return parseFloat(value);
            case 'bool':
                return Boolean(value);
            case 'string':
                return String(value);
            case 'date':
                // Create date at midnight UTC
                return new Date(value + 'T00:00:00.000Z');
            case 'datetime':
                // Datetime - parse as-is (should have Z suffix)
                return new Date(value);
            case 'time':
                // Time is stored as Date on epoch (1970-01-01) UTC
                return new Date('1970-01-01T' + value + 'Z');
            case 'decimal':
                // JS doesn't have native Decimal, skip these tests
                return value;
            default:
                return value;
        }
    }
    return spec;
}

/**
 * Recursively create JavaScript object from fixture spec.
 */
function createObject(obj) {
    if (obj && typeof obj === 'object') {
        if ('type' in obj && 'value' in obj && Object.keys(obj).length === 2) {
            return createValue(obj);
        }
        if (Array.isArray(obj)) {
            return obj.map(createObject);
        }
        const result = {};
        for (const [k, v] of Object.entries(obj)) {
            if (!k.startsWith('_')) {
                result[k] = createObject(v);
            }
        }
        return result;
    }
    return obj;
}

describe('Cross-language: typed_text', () => {
    for (const testCase of fixtures.typed_text.cases) {
        // Skip python-only tests (Decimal)
        if (testCase._python_only) {
            continue;
        }

        const value = createValue(testCase.input);
        const expected = testCase.expected;

        test(`as_typed_text(${JSON.stringify(testCase.input)}) = "${expected}"`, () => {
            const result = registry.as_typed_text(value);
            assert.strictEqual(result, expected);
        });
    }
});

describe('Cross-language: from_text', () => {
    for (const testCase of fixtures.from_text.cases) {
        // Skip python-only tests (Decimal)
        if (testCase._python_only) {
            continue;
        }

        const inputStr = testCase.input;
        const expectedSpec = testCase.expected;

        test(`from_text("${inputStr}")`, () => {
            const result = registry.from_text(inputStr);

            switch (expectedSpec.type) {
                case 'int':
                    assert.strictEqual(typeof result, 'number');
                    assert.strictEqual(result, expectedSpec.value);
                    break;
                case 'float':
                    assert.strictEqual(typeof result, 'number');
                    assert.ok(Math.abs(result - expectedSpec.value) < 0.0001);
                    break;
                case 'bool':
                    assert.strictEqual(typeof result, 'boolean');
                    assert.strictEqual(result, expectedSpec.value);
                    break;
                case 'string':
                    assert.strictEqual(typeof result, 'string');
                    assert.strictEqual(result, expectedSpec.value);
                    break;
                case 'date':
                    assert.ok(result instanceof Date);
                    assert.strictEqual(result.toISOString().slice(0, 10), expectedSpec.value);
                    break;
                case 'datetime':
                    assert.ok(result instanceof Date);
                    // Compare UTC datetime - fixture has Z suffix
                    const expectedIso = expectedSpec.value.replace('Z', '');
                    assert.strictEqual(result.toISOString().slice(0, 19), expectedIso);
                    break;
                case 'time':
                    // Time is now a Date on epoch
                    assert.ok(result instanceof Date);
                    assert.strictEqual(result.getUTCFullYear(), 1970);
                    assert.strictEqual(result.getUTCMonth(), 0);
                    assert.strictEqual(result.getUTCDate(), 1);
                    // Compare time string
                    const timeStr = `${String(result.getUTCHours()).padStart(2, '0')}:${String(result.getUTCMinutes()).padStart(2, '0')}:${String(result.getUTCSeconds()).padStart(2, '0')}`;
                    assert.strictEqual(timeStr, expectedSpec.value);
                    break;
            }
        });
    }
});

describe('Cross-language: typed_json', () => {
    for (const testCase of fixtures.typed_json.cases) {
        const name = testCase.name || 'unnamed';

        test(`as_typed_json: ${name}`, () => {
            const inputObj = createObject(testCase.input);
            const expected = testCase.expected;

            const resultStr = as_typed_json(inputObj);
            const result = JSON.parse(resultStr);

            // Deep compare
            assert.deepStrictEqual(result, expected);
        });
    }
});

describe('Cross-language: json_roundtrip', () => {
    for (const testCase of fixtures.json_roundtrip.cases) {
        const name = testCase.name || 'unnamed';

        test(`roundtrip: ${name}`, () => {
            const inputObj = createObject(testCase.input);

            const jsonStr = as_typed_json(inputObj);
            const result = from_json(jsonStr);

            // Compare structure and values
            assertEqualTyped(result, inputObj, name);
        });
    }
});

/**
 * Assert two values are equal with correct types.
 */
function assertEqualTyped(result, expected, name) {
    if (expected && typeof expected === 'object' && !Array.isArray(expected) && !(expected instanceof Date)) {
        assert.ok(typeof result === 'object' && result !== null, `${name}: expected object`);
        const expectedKeys = Object.keys(expected);
        const resultKeys = Object.keys(result);
        assert.deepStrictEqual(new Set(resultKeys), new Set(expectedKeys), `${name}: keys mismatch`);
        for (const k of expectedKeys) {
            assertEqualTyped(result[k], expected[k], `${name}.${k}`);
        }
    } else if (Array.isArray(expected)) {
        assert.ok(Array.isArray(result), `${name}: expected array`);
        assert.strictEqual(result.length, expected.length, `${name}: length mismatch`);
        for (let i = 0; i < expected.length; i++) {
            assertEqualTyped(result[i], expected[i], `${name}[${i}]`);
        }
    } else if (expected instanceof Date) {
        assert.ok(result instanceof Date, `${name}: expected Date`);
        assert.strictEqual(result.toISOString(), expected.toISOString());
    } else if (typeof expected === 'number') {
        assert.strictEqual(typeof result, 'number', `${name}: expected number`);
        assert.ok(Math.abs(result - expected) < 0.0001, `${name}: ${result} != ${expected}`);
    } else {
        assert.strictEqual(result, expected, `${name}: value mismatch`);
    }
}

File: tests/test_cross_language.py
# Copyright 2025 Softwell S.r.l.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Cross-language compatibility tests for TYTX protocol.

These tests verify that Python implementation produces output
identical to the shared fixtures that JS and TS must also match.
"""

import json
from datetime import date, datetime, time
from decimal import Decimal
from pathlib import Path

import pytest

import genro_tytx as tytx

FIXTURES_PATH = Path(__file__).parent / "fixtures" / "cross_language.json"


@pytest.fixture
def fixtures():
    """Load cross-language test fixtures."""
    with open(FIXTURES_PATH) as f:
        return json.load(f)


def _create_value(spec):
    """Create Python value from fixture spec."""
    if isinstance(spec, dict) and "type" in spec:
        type_name = spec["type"]
        value = spec["value"]
        if type_name == "int":
            return int(value)
        elif type_name == "float":
            return float(value)
        elif type_name == "bool":
            return bool(value)
        elif type_name == "string":
            return str(value)
        elif type_name == "date":
            return date.fromisoformat(value)
        elif type_name == "datetime":
            # Handle Z suffix for Python 3.10 compatibility
            if value.endswith("Z"):
                value = value[:-1] + "+00:00"
            return datetime.fromisoformat(value)
        elif type_name == "time":
            return time.fromisoformat(value)
        elif type_name == "decimal":
            return Decimal(value)
    return spec


def _create_object(obj):
    """Recursively create Python object from fixture spec."""
    if isinstance(obj, dict):
        if "type" in obj and "value" in obj and len(obj) == 2:
            return _create_value(obj)
        return {k: _create_object(v) for k, v in obj.items() if not k.startswith("_")}
    elif isinstance(obj, list):
        return [_create_object(item) for item in obj]
    return obj


class TestTypedText:
    """Test as_typed_text produces expected output."""

    def test_typed_text_cases(self, fixtures):
        """Verify as_typed_text output matches fixtures."""
        for case in fixtures["typed_text"]["cases"]:
            value = _create_value(case["input"])
            expected = case["expected"]
            result = tytx.as_typed_text(value)
            assert (
                result == expected
            ), f"as_typed_text({value!r}) = {result!r}, expected {expected!r}"


class TestFromText:
    """Test from_text parses typed strings correctly."""

    def test_from_text_cases(self, fixtures):
        """Verify from_text output matches fixtures."""
        for case in fixtures["from_text"]["cases"]:
            input_str = case["input"]
            expected_spec = case["expected"]
            result = tytx.from_text(input_str)

            expected_type = expected_spec["type"]
            expected_value = expected_spec["value"]

            if expected_type == "int":
                assert isinstance(result, int)
                assert result == expected_value
            elif expected_type == "float":
                assert isinstance(result, float)
                assert result == pytest.approx(expected_value)
            elif expected_type == "bool":
                assert isinstance(result, bool)
                assert result == expected_value
            elif expected_type == "string":
                assert isinstance(result, str)
                assert result == expected_value
            elif expected_type == "date":
                assert isinstance(result, date)
                assert result == date.fromisoformat(expected_value)
            elif expected_type == "datetime":
                assert isinstance(result, datetime)
                # Handle Z suffix for Python 3.10 compatibility
                if expected_value.endswith("Z"):
                    expected_value = expected_value[:-1] + "+00:00"
                assert result == datetime.fromisoformat(expected_value)
            elif expected_type == "time":
                assert isinstance(result, time)
                assert result == time.fromisoformat(expected_value)
            elif expected_type == "decimal":
                assert isinstance(result, Decimal)
                assert str(result) == expected_value


class TestTypedJson:
    """Test as_typed_json produces expected output."""

    def test_typed_json_cases(self, fixtures):
        """Verify as_typed_json output matches fixtures."""
        for case in fixtures["typed_json"]["cases"]:
            input_obj = _create_object(case["input"])
            expected = case["expected"]

            result_str = tytx.as_typed_json(input_obj)
            result = json.loads(result_str)

            assert result == expected, (
                f"Case '{case.get('name', 'unnamed')}': "
                f"as_typed_json({input_obj!r}) = {result!r}, expected {expected!r}"
            )


class TestJsonRoundtrip:
    """Test JSON roundtrip preserves values."""

    def test_roundtrip_cases(self, fixtures):
        """Verify from_json(as_typed_json(x)) == x."""
        for case in fixtures["json_roundtrip"]["cases"]:
            input_obj = _create_object(case["input"])

            json_str = tytx.as_typed_json(input_obj)
            result = tytx.from_json(json_str)

            # Compare with expected types
            self._assert_equal_typed(result, input_obj, case.get("name", "unnamed"))

    def _assert_equal_typed(self, result, expected, name):
        """Assert two values are equal with correct types."""
        if isinstance(expected, dict):
            assert isinstance(
                result, dict
            ), f"{name}: expected dict, got {type(result)}"
            assert set(result.keys()) == set(expected.keys()), f"{name}: keys mismatch"
            for k in expected:
                self._assert_equal_typed(result[k], expected[k], f"{name}.{k}")
        elif isinstance(expected, list):
            assert isinstance(
                result, list
            ), f"{name}: expected list, got {type(result)}"
            assert len(result) == len(expected), f"{name}: length mismatch"
            for i, (r, e) in enumerate(zip(result, expected, strict=True)):
                self._assert_equal_typed(r, e, f"{name}[{i}]")
        elif isinstance(expected, date) and not isinstance(expected, datetime):
            assert isinstance(
                result, date
            ), f"{name}: expected date, got {type(result)}"
            assert result == expected
        elif isinstance(expected, datetime):
            assert isinstance(
                result, datetime
            ), f"{name}: expected datetime, got {type(result)}"
            assert result == expected
        elif isinstance(expected, float):
            assert isinstance(
                result, (int, float)
            ), f"{name}: expected number, got {type(result)}"
            assert result == pytest.approx(expected)
        else:
            assert result == expected, f"{name}: {result!r} != {expected!r}"
